### PR TITLE
Handle error response from cloud shell and add tests

### LINKED ISSUE #4402: [Bug] MSI cloud shell the error message is not parsed from the error response.
### Library version used

NA

### .NET version

Net 6

### Scenario

ManagedIdentityClient - managed identity

### Is this a new or an existing app?

This is a new app or experiment

### Issue description and reproduction steps

The error response for cloud shell is different and needs to be handled differently to get the error message.

Currently for managed identity error response we only look at the top level for error message and error response. For cloud shell the error message is one level down in the json.

Sample error response from cloud shell when a wrong scope is passed.

```
"{\"error\":{\"code\":\"AudienceNotSupported\",\"message\":\"Audiencehttp://management.azure.comisnotasupportedMSItokenaudience.Supportedaudiences:https://management.core.windows.net/,https://management.azure.com/,https://graph.windows.net/,https://vault.azure.net,https://datalake.azure.net/,https://outlook.office365.com/,https://graph.microsoft.com/,https://batch.core.windows.net/,https://analysis.windows.net/powerbi/api,https://storage.azure.com/,https://rest.media.azure.net,https://api.loganalytics.io,https://ossrdbms-aad.database.windows.net,https://www.yammer.com,https://digitaltwins.azure.net,0b07f429-9f4b-4714-9392-cc5e8e80c8b0,822c8694-ad95-4735-9c55-256f7db2f9b4,https://dev.azuresynapse.net,https://database.windows.net,https://quantum.microsoft.com,https://iothubs.azure.net,2ff814a6-3304-4ab8-85cb-cd0e6f879c1d,https://azuredatabricks.net/,ce34e7e5-485f-4d76-964f-b3d2b16d1e4f,https://azure-devices-provisioning.net,https://managedhsm.azure.net,499b84ac-1321-427f-aa17-267ca6975798,https://api.adu.microsoft.com/,https://purview.azure.net/,6dae42f8-4368-4678-94ff-3960e28e3630\"}}"
```

### Relevant code snippets

_No response_

### Expected behavior

_No response_

### Identity provider

Microsoft Entra ID (Work and School accounts and Personal Microsoft accounts)

### Regression

_No response_

### Solution and workarounds

Look for the error message one level down in the json for cloud shell.

### PR DESCRIPTION
**Pull request by neha-bhargava**
Fixes #4402 

**Changes proposed in this request**
* Add error message parsing for cloud shell error response where the message is in nested JSON
* Add mock for error messages from different sources.

**Testing**
Updated unit tests

**Performance impact**

**Documentation**
- [ ] All relevant documentation is updated.

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs [hunk 1]
```diff
@@ -136,6 +141,36 @@ internal static string GetMessageFromErrorResponse(HttpResponse response)
...
+            StringBuilder errorMessage = new StringBuilder("[Managed Identity] ");
+
+            if (errorCode != null)
+            {
+                errorMessage.Append($"Error Code: {errorCode} ");
+            }
+
+            if (message != null)
+            {
+                errorMessage.Append($"[Managed Identity] Error Message: {message}");
```
Comment (gladjohn) [inline src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs:168]:
```suggestion
                errorMessage.Append($"Error Message: {message}");
```

File: src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs [hunk 2]
```diff
@@ -136,6 +141,36 @@ internal static string GetMessageFromErrorResponse(HttpResponse response)
             return $"[Managed Identity] Error Code: {managedIdentityErrorResponse.Error} Error Message: {managedIdentityErrorResponse.ErrorDescription}";
         }
 
+        // Try to get the error message from the nested error response in case of cloud shell.
+        private string TryGetMessageFromNestedErrorResponse(string response)
+        {
+            var json = JsonHelper.ParseIntoJsonObject(response);
```
Comment (gladjohn) [inline src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs:147]: can json be empty? if so, you can do this, 

```cs
if (JsonHelper.TryGetValue(json, "error", out var error)){}

File: src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs [hunk 3]
```diff
@@ -136,6 +141,36 @@ internal static string GetMessageFromErrorResponse(HttpResponse response)
...
+
+            JsonHelper.TryGetValue(json, "error", out var error);
+            JsonHelper.TryGetValue(JsonHelper.ToJsonObject(error), "message", out var message);
+            JsonHelper.TryGetValue(JsonHelper.ToJsonObject(error), "code", out var errorCode);
+
+            if (message == null && errorCode == null)
```
Comment (gladjohn) [inline src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs:153]: you can return early with `errorMessage` and move this condition i.e. the generic error message to the bottom.

File: src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs [hunk 4]
```diff
@@ -136,6 +141,36 @@ internal static string GetMessageFromErrorResponse(HttpResponse response)
...
+            {
+                _requestContext.Logger.Error(MsalErrorMessage.ManagedIdentityUnexpectedErrorResponse + $"Error response received from the server: {response}.");
+                return MsalErrorMessage.ManagedIdentityUnexpectedErrorResponse + $"Error response received from the server: {response}.";
```
Comment (gladjohn) [inline src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs:156]:
```cs
$"{MsalErrorMessage.ManagedIdentityUnexpectedErrorResponse}. Error response received from the server: {response}.";
```

File: src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs [hunk 5]
```diff
@@ -119,13 +119,18 @@ protected ManagedIdentityResponse GetSuccessfulResponse(HttpResponse response)
...
 
-        internal static string GetMessageFromErrorResponse(HttpResponse response)
+        internal string GetMessageFromErrorResponse(HttpResponse response)
         {
+            if (string.IsNullOrEmpty(response?.Body))
+            {
+                return MsalErrorMessage.ManagedIdentityNoResponseReceived;
+            }
+
             ManagedIdentityErrorResponse managedIdentityErrorResponse = JsonHelper.TryToDeserializeFromJson<ManagedIdentityErrorResponse>(response?.Body);
```
Comment (gladjohn) [inline src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs:129]: can managedIdentityErrorResponse Error and ErrorDescription be null ?

File: src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs [hunk 6]
```diff
@@ -136,6 +141,36 @@ internal static string GetMessageFromErrorResponse(HttpResponse response)
...
+
+            if (message == null && errorCode == null)
+            {
+                _requestContext.Logger.Error(MsalErrorMessage.ManagedIdentityUnexpectedErrorResponse + $"Error response received from the server: {response}.");
+                return MsalErrorMessage.ManagedIdentityUnexpectedErrorResponse + $"Error response received from the server: {response}.";
+            }
+
+            StringBuilder errorMessage = new StringBuilder("[Managed Identity] ");
+
+            if (errorCode != null)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs:161]: This logic where you have an error code or message isn't unit tested.

Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs]: Refactored to cover all the scenarios. Also added all kinds of expected json responses known from the server [here[(https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/nebharg/msiCloudShellFix/tests/Microsoft.Identity.Test.Common/Core/Mocks/MockHelpers.cs). Will add different error responses to unit test the error message parsing behavior.

File: src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs [hunk 7]
```diff
@@ -119,21 +119,83 @@ protected ManagedIdentityResponse GetSuccessfulResponse(HttpResponse response)
...
             if (!string.IsNullOrEmpty(managedIdentityErrorResponse.Message))
-            { 
-                return $"[Managed Identity] Error Message: {managedIdentityErrorResponse.Message} Managed Identity Correlation ID: {managedIdentityErrorResponse.CorrelationId} Use this Correlation ID for further investigation.";
+            {
+                stringBuilder.Append($"Error Message: {managedIdentityErrorResponse.Message} ");
+            }
+
+            if (!string.IsNullOrEmpty(managedIdentityErrorResponse.ErrorDescription))
+            {
+                stringBuilder.Append($"Error Message: {managedIdentityErrorResponse.ErrorDescription} ");
```
Comment (gladjohn) [inline src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs:156]:
```suggestion
                stringBuilder.Append($"Error Description: {managedIdentityErrorResponse.ErrorDescription} ");
```

**Review by bgavrilMS — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs [hunk 8]
```diff
@@ -119,21 +119,88 @@ protected ManagedIdentityResponse GetSuccessfulResponse(HttpResponse response)
...
+            }
+            catch
+            {
+                return TryGetMessageFromNestedErrorResponse(response.Body);
+            } 
+        }
+
+        private string ExtractErrorMessageFromManagedIdentityErrorResponse(ManagedIdentityErrorResponse managedIdentityErrorResponse)
+        {
+            StringBuilder stringBuilder = new StringBuilder("[Managed Identity] ");
```
Comment (gladjohn) [inline src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs:142]: consider storing this into a `ManagedIdentityPrefix` string const and checking for, 

```cs
if (stringBuilder.Length == ManagedIdentityPrefix.Length)
{
   return $"{MsalErrorMessage.ManagedIdentityUnexpectedErrorResponse}.";
}

```

**Review by gladjohn — APPROVED** [review]

**Review by gladjohn — APPROVED** [review]

**Comment by bgavrilMS** [comment]
@neha-bhargava - anything pending on this one?

**Comment by neha-bhargava** [comment]
> @neha-bhargava - anything pending on this one?

No, just the builds. Merging now

**Pull request merged**